src/Common/src/System/Globalization/FormatProvider.Number.cs

```diff
@@ -627,48 +627,73 @@ internal static unsafe void Int32ToDecChars(char* buffer, ref int index, uint va
                 }
             }
 
-            internal static unsafe char ParseFormatSpecifier(string format, out int digits)
+            internal static unsafe char ParseFormatSpecifier(ReadOnlySpan<char> format, out int digits)
             {
-                if (format != null)
+                char c = default;
+                if (format.Length > 0)
                 {
-                    fixed (char* pFormat = format)
+                    // If the format begins with a symbol, see if it's a standard format
+                    // with or without a specified number of digits.
+                    c = format[0];
+                    if ((uint)(c - 'A') <= 'Z' - 'A' ||
+                        (uint)(c - 'a') <= 'z' - 'a')
                     {
-                        int i = 0;
-                        char ch = pFormat[i];
-                        if (ch != 0)
+                        // Fast path for sole symbol, e.g. "D"
+                        if (format.Length == 1)
                         {
-                            if (((ch >= 'A') && (ch <= 'Z')) || ((ch >= 'a') && (ch <= 'z')))
+                            digits = -1;
+                            return c;
+                        }
+
+                        if (format.Length == 2)
```

Comment:
Given the new parsing logic here, do we have tests that confirm the behavior of non-conforming and strange inputs?

`D\0`
`D4\0`
`DZ`
`D\u0664` (Arabic-Indic 4)`

I'm not particularly interested in how well any of these are handled, just that they're handled the same before and after

Reply:
> Given the new parsing logic here, do we have tests that confirm the behavior of non-conforming and strange inputs?

We do for the same code in coreclr but not here.  I'll add some.